util/threadpool_imp.cc

```diff
@@ -189,10 +220,20 @@ void ThreadPoolImpl::Impl::BGThread(size_t thread_id) {
     // Wait until there is an item that is ready to run
     std::unique_lock<std::mutex> lock(mu_);
     // Stop waiting if the thread needs to do work or needs to terminate.
+    // Increase num_waiting_threads_ once this task has started waiting
+    num_waiting_threads_++;
+
+    // When not exist_all_threads and the current thread id is not the last
+    // excessive thread, it may be blocked due to 3 reasons: 1) queue is empty
+    // 2) it is the excessive thread (not the last one)
+    // 3) the number of reserved threads reaches the number of waiting threads
     while (!exit_all_threads_ && !IsLastExcessiveThread(thread_id) &&
-           (queue_.empty() || IsExcessiveThread(thread_id))) {
+           (queue_.empty() || IsExcessiveThread(thread_id) ||
+            num_waiting_threads_ == reserved_threads_)) {
```

Comment:
Sorry I'm missing the intuition. How does `num_waiting_threads_` ever reach `reserved_threads_` when queue is nonempty and the thread is not excessive? Won't it spin the CPU?

Reply:
The queue can be nonempty after reservation. Several tasks are then put in the queue. Then A thread releases fewer number of threads, at this moment, a third thread issues another reservation.

Reply:
Typically, earlier buggy implementation failed at `ASSERT_EQ(1U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));' Because at that point, thread with id 0 has been waiting and its id is not excessive, this will make thread 0 execute the incoming task 3, instead of being blocked. Adding the `num_waiting_thread_ == reserved_threads_` can ensure thread 0 is blocked and Task 3 is in the queue.

Reply:
updated with replacing the condition with `num_waiting_threads_  <= reserved_threads_`.

Reply:
Will need to catch up on this tomorrow but I am thinking (1) whether num_waiting_threads_  < num_reserved_threads is legitimate and (2) if it is, whether we should revise the design to eliminate that cuz it’s pretty counter-intuitive given the Reserve() algorithm we are pursuing.

Reply:
For the current design, it is very possible that num_waiting_threads < num_reserved_threads because of the excessive case (andrew mentioned earlier). I will replace num_waiting_threads - num_reserved_threads with max(num_waiting_threads - num_reserved_threads, 0) in the next commit

Reply:
Right - `SetBackgroundThreadsInternal` can be called at any point and we can't terminate any running threads or reserved threads but num_waiting_threads.

It's make things less obvious cuz people or at least I'd easily think "We just do now allow more threads to be reserved when that occurs." should NOT happen (cuz reserved threads are like running threads, just running elsewhere). I might think of failing or delaying `SetBackgroundThreadsInternal` in this case. 

However, this might be too big a follow-up work for too infrequent use case.  So I'm good with adding a brief note and explanation
about "why there would ever be num_waiting_threads < num_reserved_threads" around the `num_waiting_threads` declaration" around the variable "num_waiting_threads" declaration as I suggested in another comment.

Reply:
As you can see from the origin implementation, in some cases, the number of running threads could be larger than the thread pool. In other words, the total_threads_limit_ is not a strict constraint, and that's why we allow that some extra reserved threads to run even though it is larger than the number of waiting threads.

Reply:
Got you!